### ISSUE TITLE
Regression: False positive where variable with type of TypeVar does not respect the fact that it's been narrowed

### ISSUE THREAD

**Issue by jshumphrey** [microsoft/pyright#7302]
I'm running Pyright v.1.1.348 as part of Pylance v.2024.2.2 in Visual Studio Code. I updated VS Code this morning, and a portion of code that previously did not display any typing errors is now reporting errors. 

To the best of my understanding, the code is valid, I have not been able to find another open or closed issue reporting the same problem, and I do not believe that any updates to Pyright beyond 1.1.348 pertain to this precise problem, so I'm reporting this as an issue.

I can't post the exact code in question, but here's a minimal example that reports the same errors. (For the record, the purpose of using a TypeVar here is to ensure that the output type matches the input type.)
```py
from __future__ import annotations
import typing

StringOrNone = typing.TypeVar("StringOrNone", str, None)
char_replacements: dict[str, str]  # Things like replacing "MS Word quotes" with regular quotes

def replace_bad_chars(input_string: StringOrNone) -> StringOrNone:
    if input_string is None:
        return input_string

    for bad_char in set(input_string) & set(char_replacements.keys()):
        # The following line now shows a typing error, and did not do so before the update
        input_string = input_string.replace(bad_char, char_replacements[bad_char])
    
    return input_string
```
Previously, the line beginning with `input_string = ` did not show a typing error. However, I now get the following two errors on this line: 
- `Expression of type "str* | Unknown" cannot be assigned to declared type "StringOrNone@_replace_bad_chars"`
- `"replace" is not a known member of "None"`

This... doesn't seem to make sense to me. By the time that we're doing character replacements, we know that `input_string` is `str`, not `None`; if it was `None`, we'd have returned immediately. On top of this, we know that we're not turning `input_string` into something other than a `str`, because `char_replacements` is a method of `str` that, itself, returns `str`. I don't understand where `Unknown` is able to enter the picture here.

Please let me know of any additional diagnostics or information I can provide, and I'll respond as soon as possible. Thank you!

**Comment by erictraut** [comment]
I don't think there have been any recent changes to pyright that would impact this behavior. Indeed, I've gone back as far as 1.1.344, and I see the same behavior with your code sample. Can you validate that you're seeing different behavior if you revert back to 1.1.348 or older? 

Note: I needed to make a small adjustment to your code sample because the variable `char_replacements` is never assigned a value in your code above, so its type is `Unbound`.

**Comment by jshumphrey** [comment]
I do not have the Pyright extension installed directly; it's bundled in with the Pylance extension in VS Code. Downgrading the Pylance extension to v2023.12.1 removes the errors; [this corresponds to Pyright v1.1.336](https://github.com/microsoft/pyright/commit/a78e2b10).

Here are snippets from the Output (Python Language Server) tab in VS Code for the various versions I've tried. These are the current version and the previous two versions of Pylance available in VS Code.

**Pylance v2024.2.2 (current version); displays error messages**
```
Pylance async client (2024.2.2) started with python extension (2024.0.1)
Pylance language server 2024.2.2 (pyright version 1.1.348, commit cfb1de0c) starting
```

**Pylance v2024.2.1; displays error messages**
```
Pylance async client (2024.2.1) started with python extension (2024.0.1)
Pylance language server 2024.2.1 (pyright version 1.1.348, commit cfb1de0c) starting
```

**Pylance v2023.12.1; does NOT display error messages**
```
Pylance async client (2023.12.1) started with python extension (2024.0.1)
Pylance language server 2023.12.1 (pyright a78e2b10) starting
```

**Comment by jshumphrey** [comment]
[Testing in the Pyright Playground](https://pyright-play.net/?pyrightVersion=1.1.344&strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoHfQqJAT1y1IHMWLAMpIQffgHkQAOTpooAXk48JAOgAqPNADViIABQAiUeIHS5pNEYA0UemLuW0AShYBjABb6Y6XABtidzQINFIkegAuKAATLHckAG0HEDsUgF0lKABvI2IjaKMAQSMAXygoAGIoDU8Jeih-LABrBT9A9wkoIwBZYSgAdQIYqABHBEo0eiMoAHcsJE8odH4EQJAxiaQpoWYYtGBltACgjAAjYhiYL316Az5cZBgUiWjTCQt5FygAWgA%2BKDvcyyeSRFgVCpYQ4PJ4vATYBrOMHMCGoo4oECkbCkR5IZ5iCS7VHAAhQC5XG4bPj2NBIe442EEgTfABkNLplN8xw6ITCETUrS4dxcLmRaIh1VqChJ-n8YHm8KaVigFFm9k88oaxBUvHhaBA4FSJFIIziIwoRBiYHs1rOBwICkWCgQuBi1DQ4PF2Nx%2BLM-CyMLxcP4anapwM5Ou3iNnLDwVC4XoiUjlPS309FQz6NQWMDvokQA) shows that the first version of Pyright to report an error on this code is Pyright v1.1.344.

**Comment by erictraut** [comment]
OK, thanks for the additional info.

I've narrowed down the change to [this commit](https://github.com/microsoft/pyright/commit/8c178ce306b7758c607a713edc0c25a40cdb3503).

**Comment by erictraut** [comment]
Thanks for the bug report. It turns out that the change in this commit uncovered a different long-standing bug in pyright. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
